models/DataObject/AbstractObject.php

```diff
@@ -468,30 +472,30 @@ public function getSiblings($objectTypes = [self::OBJECT_TYPE_OBJECT, self::OBJE
             $list->setOrderKey('o_key');
             $list->setObjectTypes($objectTypes);
             $list->setOrder('asc');
-            $this->o_siblings = $list->load();
+            $this->o_siblings[$cacheKey] = $list->load();
         }
 
-        return $this->o_siblings;
+        return $this->o_siblings[$cacheKey];
     }
 
     /**
      * Returns true if the object has at least one sibling
      *
      * @param array $objectTypes
+     * @param bool $unpublished
      *
      * @return bool
      */
-    public function hasSiblings($objectTypes = [self::OBJECT_TYPE_OBJECT, self::OBJECT_TYPE_FOLDER])
+    public function hasSiblings($objectTypes = [self::OBJECT_TYPE_OBJECT, self::OBJECT_TYPE_FOLDER], $unpublished = false)
     {
-        if (is_bool($this->o_hasSiblings)) {
-            if (($this->o_hasSiblings and empty($this->o_siblings)) or (!$this->o_hasSiblings and !empty($this->o_siblings))) {
-                return $this->getDao()->hasSiblings($objectTypes);
-            } else {
-                return $this->o_hasSiblings;
+        if (isset($this->o_hasSiblings[$unpublished])) {
```

Comment:
Here we also need to consider the `$objectTypes` for caching, use `getListingCacheKey()` here as well.